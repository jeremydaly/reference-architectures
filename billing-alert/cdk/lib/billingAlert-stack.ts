import * as cdk from '@aws-cdk/core';
import * as cloudwatch from "@aws-cdk/aws-cloudwatch";
import * as cloudwatch_actions from "@aws-cdk/aws-cloudwatch-actions";
import * as sns from "@aws-cdk/aws-sns";
import * as sns_subscriptions from "@aws-cdk/aws-sns-subscriptions";

export interface BillingAlertStackProps extends cdk.StackProps {
    alertThresholdInDollars: number,
    email: string
}

export class BillingAlertStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props: BillingAlertStackProps) {
        super(scope, id, props);

        const alarm = new cloudwatch.Alarm(this, "SpendingAlarm", {
            metric: new cloudwatch.Metric({
                namespace: "AWS/Billing",
                metricName: "EstimatedCharges",
                dimensions: {"Currency": "USD"},
                period: cdk.Duration.hours(6), // 21600,
                statistic: "Maximum"
            }),

            alarmDescription: `Alarm if AWS spending is over $${props.alertThresholdInDollars}`,
            comparisonOperator: cloudwatch.ComparisonOperator.GREATER_THAN_THRESHOLD,
            evaluationPeriods: 1,
            threshold: props.alertThresholdInDollars,
        })

        const topic = new sns.Topic(this, "SpendingTopic")
        topic.addSubscription(new sns_subscriptions.EmailSubscription(props.email))

        const snsAction = new cloudwatch_actions.SnsAction(topic)
        alarm.addAlarmAction(snsAction)
        alarm.addInsufficientDataAction(snsAction)
    }
}
