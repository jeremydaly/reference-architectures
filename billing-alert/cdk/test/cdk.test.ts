import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as Cdk from '../lib/billingAlert-stack';

test('stack creates alert', () => {
    const app = new cdk.App();
    // WHEN
    const stack = new Cdk.BillingAlertStack(app, 'BillingAlertStack', {
        alertThresholdInDollars: 100,
        email: "milton@initech.com"
    });
    // THEN
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
