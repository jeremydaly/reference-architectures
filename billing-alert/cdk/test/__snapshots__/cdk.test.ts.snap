// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stack creates alert 1`] = `
Object {
  "Resources": Object {
    "SpendingAlarm736B4791": Object {
      "Properties": Object {
        "AlarmActions": Array [
          Object {
            "Ref": "SpendingTopic7406B88C",
          },
        ],
        "AlarmDescription": "Alarm if AWS spending is over $100",
        "ComparisonOperator": "GreaterThanThreshold",
        "Dimensions": Array [
          Object {
            "Name": "Currency",
            "Value": "USD",
          },
        ],
        "EvaluationPeriods": 1,
        "InsufficientDataActions": Array [
          Object {
            "Ref": "SpendingTopic7406B88C",
          },
        ],
        "MetricName": "EstimatedCharges",
        "Namespace": "AWS/Billing",
        "Period": 21600,
        "Statistic": "Maximum",
        "Threshold": 100,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "SpendingTopic7406B88C": Object {
      "Type": "AWS::SNS::Topic",
    },
    "SpendingTopicmiltoninitechcom8303C3F6": Object {
      "Properties": Object {
        "Endpoint": "milton@initech.com",
        "Protocol": "email",
        "TopicArn": Object {
          "Ref": "SpendingTopic7406B88C",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
  },
}
`;
